### PR TITLE
Release message block chains after each get_*_value(s) call in DynamicData

### PR DESCRIPTION
**Pull request by sonndinh**

### CODE REVIEW

File: dds/DCPS/XTypes/DynamicData.h [hunk 1]
```diff
@@ -154,7 +154,30 @@ class OpenDDS_Dcps_Export DynamicData {
 
   bool check_xcdr1_mutable(const DynamicType_rch& dt);
 
+  typedef OPENDDS_VECTOR(ACE_Message_Block*) IntermediateChains;
+  const IntermediateChains& get_intermediate_chains() { return chains_to_release; }
```
Comment (mitza-oci) [inline dds/DCPS/XTypes/DynamicData.h:158]: Could these be private?

Comment (mitza-oci) [inline dds/DCPS/XTypes/DynamicData.h:158]:
```suggestion
  typedef OPENDDS_VECTOR(ACE_Message_Block*) IntermediateChains;
  const IntermediateChains& get_intermediate_chains() const { return chains_to_release; }
```

File: dds/DCPS/XTypes/DynamicData.h [hunk 2]
```diff
@@ -154,7 +154,30 @@ class OpenDDS_Dcps_Export DynamicData {
...
+
 private:
+
+  class ScopedChainManager {
+  public:
+    ScopedChainManager(DynamicData& dd)
```
Comment (mitza-oci) [inline dds/DCPS/XTypes/DynamicData.h:164]:
```suggestion
    explicit ScopedChainManager(DynamicData& dd)
```

**Review by mitza-oci — CHANGES_REQUESTED** [review]

**Review by mitza-oci — APPROVED** [review]
Does this use case work?  I think it does, but just want to verify.  If it's not already in tests we may want to add it.
```
DynamicData d2;
{
  DynamicData d1(init_for_overall_data_sample);
  d1.get_complex_value(d2, id_nested);
}
// d2 is still valid
```

**Comment by sonndinh** [comment]
> 

Yes, it should work because `d2` is independent of `d1`. I'll add a test case for that.

**Pull request merged**